apollo-router/src/plugins/record_replay/record.rs

```diff
@@ -0,0 +1,350 @@
+use std::collections::HashMap;
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use http::HeaderMap;
+use http::HeaderValue;
+use serde_json::json;
+use tokio::fs;
+use tower::BoxError;
+use tower::ServiceBuilder;
+use tower::ServiceExt as TowerServiceExt;
+
+use super::after_complete::AfterCompleteStreamExt;
+use super::recording::RequestDetails;
+use super::recording::ResponseDetails;
+use super::recording::Subgraph;
+use super::recording::Subgraphs;
+use crate::context::Context;
+use crate::layers::ServiceBuilderExt;
+use crate::plugin::Plugin;
+use crate::plugin::PluginInit;
+use crate::services::execution;
+use crate::services::router;
+use crate::services::subgraph;
+use crate::services::supergraph;
+use crate::spec::query::Query;
+use crate::spec::Schema;
+use crate::Configuration;
+
+const RECORD_HEADER: &str = "x-apollo-router-record";
+const RECORD: &str = "record";
+const CLIENT_REQUEST: &str = "client_request";
+const CLIENT_RESPONSE: &str = "client_response";
+const QUERY_PLAN: &str = "query_plan";
+const SUBGRAPHS: &str = "subgraphs";
+
+/// Request recording configuration.
+#[derive(Debug, Clone, serde::Deserialize, schemars::JsonSchema)]
+#[serde(deny_unknown_fields)]
+struct RecordConfig {
+    /// The recording plugin is disabled by default.
+    enabled: bool,
+    /// The path to the directory where recordings will be stored. Defaults to
+    /// the current working directory.
+    #[serde(default = "default_storage_path")]
+    storage_path: String,
```

Comment:
if this is made an option, then the snapshot will not have different results per machine